src/properties.ts

```diff
@@ -0,0 +1,319 @@
+      packageJsonParams['sonar.links.issue'] = pkg.bugs.url;
+    }
+    if (pkg.repository?.url) {
+      packageJsonParams['sonar.links.scm'] = pkg.repository.url;
+    }
+
+    const potentialCoverageDirs = [
+      // jest coverage output directory
+      // See: http://facebook.github.io/jest/docs/en/configuration.html#coveragedirectory-string
+      pkg['nyc']?.['report-dir'],
+      // nyc coverage output directory
+      // See: https://github.com/istanbuljs/nyc#configuring-nyc
+      pkg['jest']?.['coverageDirectory'],
+    ]
+      .filter(Boolean)
+      .concat(
+        // default coverage output directory
+        'coverage',
+      );
+    const uniqueCoverageDirs = Array.from(new Set(potentialCoverageDirs));
+    packageJsonParams['sonar.exclusions'] = sonarBaseExclusions;
+    for (const lcovReportDir of uniqueCoverageDirs) {
+      const lcovReportPath = path.posix.join(lcovReportDir, 'lcov.info');
+      if (fileExistsInProjectSync(lcovReportPath)) {
+        packageJsonParams['sonar.exclusions'] +=
+          (packageJsonParams['sonar.exclusions'].length > 0 ? ',' : '') +
+          path.posix.join(lcovReportDir, '**');
+        // https://docs.sonarqube.org/display/PLUG/JavaScript+Coverage+Results+Import
+        packageJsonParams['sonar.javascript.lcov.reportPaths'] = lcovReportPath;
+        // TODO: use Generic Test Data to remove dependence of SonarJS, it is need transformation lcov to sonar generic coverage format
+      }
+    }
+
+    if (dependenceExists('mocha-sonarqube-reporter') && fileExistsInProjectSync('xunit.xml')) {
+      // https://docs.sonarqube.org/display/SONAR/Generic+Test+Data
+      packageJsonParams['sonar.testExecutionReportPaths'] = 'xunit.xml';
+    }
+    // TODO: use `glob` to lookup xunit format files and transformation to sonar generic report format
+  }
+  return packageJsonParams;
+}
+
+/**
+ * Convert CLI args into scanner properties.
+ */
+function getCommandLineProperties(cliArgs?: string[]): ScannerProperties {
+  if (!cliArgs || cliArgs.length === 0) {
+    return {};
+  }
+
+  // Parse CLI args (eg: -Dsonar.token=xxx)
+  const properties: ScannerProperties = {};
+  for (const arg of cliArgs) {
+    if (!arg.startsWith('-D')) {
+      continue;
+    }
+    const [key, value] = arg.substring(2).split('=');
+    properties[key] = value;
+  }
+
+  return properties;
+}
+
+/**
+ * Parse properties stored in sonar project properties file, if it exists.
+ */
+function getSonarFileProperties(projectBaseDir: string): ScannerProperties {
+  // Read sonar project properties file in project base dir
+  try {
+    const sonarPropertiesFile = path.join(projectBaseDir, SONAR_PROJECT_FILENAME);
+    const properties: ScannerProperties = {};
+    const data = fs.readFileSync(sonarPropertiesFile).toString();
+    const lines = data.split('\n');
+    for (const line of lines) {
+      const trimmedLine = line.trim();
+      if (trimmedLine.length === 0 || trimmedLine.startsWith('#')) {
+        continue;
```

Comment:
Agreed, the `trimmedLine.startsWith('#')` branch is already [covered](https://github.com/SonarSource/sonar-scanner-npm/pull/111/files#diff-3ce9a1a070fb68ac9ab3a68213e1d2a7188376dc6084c9dfce860d21cfe8fe4cR4)/[tested](https://github.com/SonarSource/sonar-scanner-npm/pull/111/files#diff-e5785e221c4617e0f634b491d9609a32a89b1b0591ec00bd286c38c1e356800bR42) but not `trimmedLine.length === 0`, I added an empty line to test it

Reply:
I meant more asserting explicitly this condition,

e.g.  `expect(properties['sonar.token']).not.toEqual('ignore-this');`

because if I comment out this line, the `sonar.token` WILL be evaluated to `ignore-this`.

Yes, it overlaps with the general assertion on the values of the properties. But IMO is more readable that we are explicitly expecting it not to be the commented value (without it, it's not clear that we explicitly have code that ignores comments. 

WDYT?

Reply:
sounds good to me, I'll change the test